internal/store/indexer.go

```diff
@@ -0,0 +1,151 @@
+package store
+
+import (
+	"fmt"
+
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	IndexNameReferrer = "referrer"
+	IndexNameReferent = "referent"
+)
+
+type ObjectReference struct {
+	Referrer client.Object
+	Referent client.Object
+}
+
+// CacheIndexers implements a reference cache to store reference relationship between k8s objects
+// provided by cache.Indexer. It could do CRUD on reference records when referrer and referent are
+// both provided. It can also list reference records by referrer or by referent.
+type CacheIndexers struct {
+	indexer cache.Indexer
+}
+
+func NewCacheIndexers() CacheIndexers {
+	return CacheIndexers{
+		indexer: cache.NewIndexer(ObjectReferenceKeyFunc,
+			cache.Indexers{
+				IndexNameReferrer: ObjectReferenceIndexerReferrer,
+				IndexNameReferent: ObjectReferenceIndexerReferent,
+			}),
+	}
+}
+
+// ObjectReferenceKeyFunc is the function to trasfer the reference relataioship
+// between k8s objects. The key is transferred to the following format:
+// group/version,Kind=kind/namespace/name:group/version,Kind=kind/namespace/name
+// The part before : is from referrer, and after : is from referent.
+func ObjectReferenceKeyFunc(obj interface{}) (string, error) {
+	ref, ok := obj.(*ObjectReference)
+	if !ok {
+		return "", fmt.Errorf("object is type %T, not ObjectReference", obj)
+	}
+	referrerKey := fmt.Sprintf("%s/%s/%s",
+		ref.Referrer.GetObjectKind().GroupVersionKind().String(),
+		ref.Referrer.GetNamespace(), ref.Referrer.GetName(),
+	)
+	referentKey := fmt.Sprintf("%s/%s/%s",
+		ref.Referent.GetObjectKind().GroupVersionKind().String(),
+		ref.Referent.GetNamespace(), ref.Referent.GetName(),
+	)
```

Comment:
Do you think we could extract these as key functions, e.g. `referrerKey(obj client.Object) string`? They are duplicated here and there and it might be easy to mess up the order of arguments for example.

Reply:
extracted to a `objectKeyFunc` to make key for an k8s object.